File: library_zap/src/main/java/com/dolin/zap/impl/ZapLifecycle.java
package com.dolin.zap.impl;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;

/**
 * @author #Suyghur.
 * Created on 2021/05/19
 */
public class ZapLifecycle {

    private static volatile ZapLifecycle mInstance = null;

    private Application application = null;


    private final Application.ActivityLifecycleCallbacks callback = new Application.ActivityLifecycleCallbacks() {
        @Override
        public void onActivityCreated(Activity activity, Bundle savedInstanceState) {

        }

        @Override
        public void onActivityStarted(Activity activity) {

        }

        @Override
        public void onActivityResumed(Activity activity) {

        }

        @Override
        public void onActivityPaused(Activity activity) {
            ZapRecord.getInstance().asyncFlush();
        }

        @Override
        public void onActivityStopped(Activity activity) {

        }

        @Override
        public void onActivitySaveInstanceState(Activity activity, Bundle outState) {

        }

        @Override
        public void onActivityDestroyed(Activity activity) {

        }
    };


    private ZapLifecycle() {

    }

    public static ZapLifecycle getInstance() {
        if (mInstance == null) {
            synchronized (ZapLifecycle.class) {
                if (mInstance == null) {
                    mInstance = new ZapLifecycle();
                }
            }
        }
        return mInstance;
    }

    public void registerZapLifeCallback(Application application) {
        this.application = application;
        application.registerActivityLifecycleCallbacks(callback);
    }

    public void unregisterZapLifeCallback() {
        if (application != null) {
            application.unregisterActivityLifecycleCallbacks(callback);
            this.application = null;
        }
    }
}

File: library_zap/src/main/java/com/suyghur/dolin/zap/impl/ZapLifecycle.kt
package com.suyghur.dolin.zap.impl

import android.app.Activity
import android.app.Application
import android.os.Bundle

/**
 * @author #Suyghur.
 * Created on 4/8/21
 */
object ZapLifecycle {

    fun registerZapLifeCallback(application: Application, record2MMap: Record2MMap) {
        application.registerActivityLifecycleCallbacks(object : Application.ActivityLifecycleCallbacks {
            override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
            }

            override fun onActivityStarted(activity: Activity) {
            }

            override fun onActivityResumed(activity: Activity) {
            }

            override fun onActivityPaused(activity: Activity) {
                record2MMap.asyncFlush()
            }

            override fun onActivityStopped(activity: Activity) {
            }

            override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {
                record2MMap.asyncFlush()
            }

            override fun onActivityDestroyed(activity: Activity) {
                record2MMap.asyncFlush()
            }

        })
    }

//    private fun initFragmentLifeCallback(): FragmentManager.FragmentLifecycleCallbacks {
//        return object : FragmentManager.FragmentLifecycleCallbacks() {
//            override fun onFragmentCreated(fm: FragmentManager, f: Fragment, savedInstanceState: Bundle?) {
//                super.onFragmentCreated(fm, f, savedInstanceState)
//            }
//
//            override fun onFragmentResumed(fm: FragmentManager, f: Fragment) {
//                super.onFragmentResumed(fm, f)
//            }
//
//        }
//    }
}